core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -570,6 +423,190 @@ class TiBatchWrite(@transient val df: DataFrame,
+          // extract index key from keyWithDupInfo
+          val indexKeys = data.indexKeys.map { key =>
+            key.key
+          }
+          indexKeys.map { key =>
+            val cdo = new CodecDataOutput()
+            cdo.writeLong(handle)
+            (key, cdo.toBytes)
+          }
+      }
+      .flatMap(identity)
+
+    rowKeyAllocatedRdd.union(indexKeyAllocatedRdd)
+  }
+
+  private def checkConflictWithInsertKeys(
+    rdd: RDD[toBeCheckedRow]
+  ) = {
+    if (handleCol != null) {
+      // data is conflict with TiKV
+      val handleConflict = !rdd
+        .filter {
+          // check handle column and unique indices
+          data =>
+            data.handleKey.oldRow != null
+        }
+        .isEmpty()
+
+      if (handleConflict) {
+        throw new TiBatchWriteException("data to be inserted is conflict on primary key")
+      }
+
+      val handleDuplicated = rdd
+        .map { row =>
+          (row.handleKey, 0)
+        }
+        .groupByKey()
+        .flatMap {
+          case (_, iterable) =>
+            if (iterable.size > 1) {
+              Some(iterable.size)
+            } else {
+              None
+            }
+        }
+
+      if (!handleDuplicated.isEmpty) {
+        throw new TiBatchWriteException("data to be inserted is conflict on primary key")
+      }
+    }
+
+    if (uniqueIndices.nonEmpty) {
+      val handleConflict = !rdd
+        .filter {
+          // check handle column and unique indices
+          data =>
+            data.indexKeys.exists(p => p.oldHandle != -1)
+        }
+        .isEmpty()
+
+      if (handleConflict) {
+        throw new TiBatchWriteException("data to be inserted is conflict on index key")
+      }
+      val indexConflict = rdd
+        .flatMap { row =>
+          row.indexKeys.map { key =>
+            (key, row)
+          }
+        }
+        .groupByKey()
+        .flatMap {
+          case (_, iterable) =>
+            if (iterable.size > 1) {
+              Some(iterable.size)
+            } else {
+              None
+            }
+        }
+
+      if (!indexConflict.isEmpty) {
+        throw new TiBatchWriteException("data to be inserted is conflict on unique index")
+      }
+    }
+
+    rdd
+  }
+
+  private def getKeysNeedCheck(
+    rdd: RDD[TiRow]
+  ) = {
+    val handleCol = if (tiTableInfo.isPkHandle) {
+      tiTableInfo.getPrimaryKeyColumn
+    } else {
+      null
+    }
+
+    //1. check is there any conflicts in data to be inserted
+    val cols = tiTableInfo.getColumns.asScala.flatMap { col =>
+      if (!col.canSkip(tiTableInfo.isPkHandle)) {
+        Some(col)
+      } else {
+        None
+      }
+    }
+
+    // TODO: replace get with batch get. It cannot be done by now since batch get api is not being
+    //  implemented correctly.
+    rdd.map { row =>
+      val handleInfo = if (handleCol != null) {
+        val handle =
+          row.get(handleCol.getOffset, handleCol.getType).asInstanceOf[Number].longValue()
```

Comment:
think about the case `table 5 columns with auto increase col & insert rdd with 4 columns`?